boa/src/object/gcobject.rs

```diff
@@ -193,19 +193,41 @@ impl GcObject {
                             Value::undefined(),
                         );
 
-                        // Add arguments object
-                        let arguments_obj = create_unmapped_arguments_object(args);
-                        local_env.borrow_mut().create_mutable_binding(
-                            "arguments".to_string(),
-                            false,
-                            true,
-                            context,
-                        )?;
-                        local_env.borrow_mut().initialize_binding(
-                            "arguments",
-                            arguments_obj,
-                            context,
-                        )?;
+                        let mut arguments_in_parameter_names = false;
+
+                        for param in params.iter() {
+                            has_parameter_expressions =
+                                has_parameter_expressions || param.init().is_some();
+                            arguments_in_parameter_names =
+                                arguments_in_parameter_names || param.name() == "arguments";
+                        }
+
+                        // An arguments object is not added in the following situations:
+                        // - In arrow functions (10.2.11.16)
+                        // - If parameter list contains `arguments` (10.2.11.17)
+                        // - If there are no default parameters and lexical names or function names contain `arguments` (10.2.11.18)
+                        //
+                        // https://tc39.es/ecma262/#sec-functiondeclarationinstantiation
+                        if !flags.is_lexical_this_mode()
+                            && !arguments_in_parameter_names
+                            && (has_parameter_expressions
+                                || (!body.lexically_declared_names().contains("arguments")
+                                    && !body.function_declared_names().contains("arguments")))
+                        {
```

Comment:
I think it might be a good idea to add something along the lines of "this if statement checks that none of the 3 conditions holds, and if so we add the arguments object", right now it looks a bit confusing to me that the comment describes ***when not to do something*** and then the code checks ***if we should do it***.
It's a complex expression that I think justifies commenting thoroughly.

Reply:
The spec defines it as "when not to do something" and the comment was based on that. Updated to represent the code instead.